### PR TITLE
Draft for SQL Server identity fluent API in model snapshot

### PR DESCRIPTION
**Pull request by roji**
@bricelam this is some quick draft work on #16922:

* Add new API to IAnnotationCodeGenerator to receive all annotations, allowing generating a single fluent API call for several annotations. Fully backwards-compatible.
* Implement handling for seed/increment via the new API
* Plumb IAnnotationCodeGenerator into CSharpSnapshotGenerator and use it to generate fluent API calls.
* Note: this currently removes annotations that are identified as by-convention by IAnnotationCodeGenerator from the context snapshopt - I'm guessing that's probably not good (since conventions may change across EF Core versions). If that's the case we should simply split IAnnotationCodeGenerator API - one for removing by-convention annotations, and another for generating fluent API calls.
* If we go down this route, we can just remove all the older IsHandledByConvention/GenerateFluentApi methods from the interface (but not from the base class), since we no longer call them directly.

If this looks like a good direction, let me know and I'll continue.

### CODE REVIEW

File: src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs [hunk 1]
```diff
@@ -524,6 +524,8 @@ protected virtual void GeneratePropertyAnnotations([NotNull] IProperty property,
 
             var annotations = property.GetAnnotations().ToList();
 
+            // First, handle built-in annotations
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Migrations/Design/CSharpSnapshotGenerator.cs:527]:
```suggestion
            // First, handle non-provider-specific annotations
```

**Comment by AndriySvyryd** [comment]
> Note: this currently removes annotations that are identified as by-convention by IAnnotationCodeGenerator from the context snapshopt - I'm guessing that's probably not good (since conventions may change across EF Core versions). If that's the case we should simply split IAnnotationCodeGenerator API - one for removing by-convention annotations, and another for generating fluent API calls.

Yes, we should do this. Otherwise it's looking good (though more elaborate than I imagined it)

**Comment by AndriySvyryd** [comment]
By the way this would also fix https://github.com/dotnet/efcore/issues/6546 Don't know why we have two issues for this

**Comment by bricelam** [comment]
+1 Looks good. Just need a way of generating by-convention annotations in the snapshot. Maybe a bool in HandleAnnotations()?

**Comment by roji** [comment]
@bricelam

> Just need a way of generating by-convention annotations in the snapshot. Maybe a bool in HandleAnnotations()?

I actually thought of just exposing two APIs - one which removes annotations which are by convention (called when scaffolding but not when generating the snapshot), and one which removes annotations for which a fluent API exists (and returns the fluent API calls).

@AndriySvyryd 

> Otherwise it's looking good (though more elaborate than I imagined it)

Did I over-engineer? Any ideas for simplifying?

>  should also consider using IAnnotationCodeGenerator to remove the annotations that are never generated

Will do this too.

**Comment by AndriySvyryd** [comment]
> Did I over-engineer? Any ideas for simplifying?

My initial design was just passing the string builder to the provider, but your way is better.

**Comment by roji** [comment]
Superceded by #21329

**Pull request closed without merging**